nemo/collections/nlp/data/language_modeling/megatron/base_prompt_learning_dataset.py

```diff
@@ -67,7 +68,52 @@ def __init__(
                 dataset = open(path, 'r', encoding='utf-8')
-                self.load_data(dataset)
+                dataset_examples = self.load_data(dataset)
+                self.examples.extend(dataset_examples)
+        elif (isinstance(datasets[0], omegaconf.ListConfig) or isinstance(datasets[0], list)):
+            # Dataset is a list of tuples with the first element being the probability of sampling from the dataset
+            # This code repeates the smaller datasets to approximately match the target probabilities
+            total_examples = 0
+            dataset_lengths = []
+            target_probs = []
+            datasets_examples_list = []
+            for prob_and_path in datasets:
+                prob = prob_and_path[0]
+                path = prob_and_path[1]
+                dataset = open(path, 'r', encoding='utf-8')
+                dataset_examples = self.load_data(dataset)
+                datasets_examples_list.append(dataset_examples)
+                dataset_lengths.append(len(dataset_examples))
+                total_examples += len(dataset_examples)
+                target_probs.append(prob)
+            
+            # Normalize the target probs
+            target_probs = [prob / sum(target_probs) for prob in target_probs]
+            current_probs = [dataset_lengths[i] / total_examples for i in range(len(dataset_lengths))]
+            
+            # Increase number of examples needed without reducing the larger datasets with low target probs
+            new_total_examples = total_examples
+            for dataset_idx in range(len(datasets)):
+                if target_probs[dataset_idx] < current_probs[dataset_idx]:
+                    target_total_examples = int(dataset_lengths[dataset_idx] / target_probs[dataset_idx])
+                    new_total_examples = max(new_total_examples, target_total_examples)
+            
+            final_total_examples = 0
+            final_dataset_lengths = []
+            for dataset_idx in range(len(datasets)):
+                num_samples_required = int(new_total_examples * target_probs[dataset_idx])
+                num_repeat = max(int(round(num_samples_required // dataset_lengths[dataset_idx])), 1) # At least 1 repeat
+                print("dataset idx", dataset_idx, "num_repeat", num_repeat)
+                dataset_examples_repeated = datasets_examples_list[dataset_idx] * num_repeat
+                final_dataset_lengths.append(len(dataset_examples_repeated))
+                final_total_examples += len(dataset_examples_repeated)
+                self.examples.extend(dataset_examples_repeated)
+            
+            final_probs = [final_dataset_lengths[i] / final_total_examples for i in range(len(final_dataset_lengths))]
+            print("Target probs: ", target_probs)
+            print("Final probs: ", final_probs)
+            print("Initial total examples: ", total_examples)
+            print("Final total examples: ", final_total_examples)
```

Comment:
Can you use logging instead of print?

Reply:
I have updated the PR changing the prints to logging